sesman/sessionrecord.c

```diff
@@ -0,0 +1,131 @@
+/**
+ * xrdp: A Remote Desktop Protocol server.
+ *
+ * Copyright (C) Emmanuel Blindauer 2017
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ *
+ * @file sessionrecord.c
+ * @brief utmp/wtmp handling code
+ * Idea: Only implement actual utmp, i.e. utmpx for 99%.
+ *       See http://80386.nl/unix/utmpx/
+ */
+
+#if defined(HAVE_CONFIG_H)
+#include <config_ac.h>
+#endif
+
+#include <paths.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/time.h>
+#include <unistd.h>
+
+#include "log.h"
+#include "os_calls.h"
+#include "sessionrecord.h"
+
+/*
+ * Prepare the utmp/ struct and write it.
+ * this can handle login and logout at once with the 'state' parameter
+ */
+
+int
+add_xtmp_entry(int pid, const char *line, const char *user, const char *rhostname, short state)
+{
+    _utmp ut;
+    struct timeval tv;
+    char *hostname = 0;
+
+    /* The string rhostname containt too much data, only get the ip
+     * the format is
+     * "2001:123:12:1234:1234:1234:1234:1234:53194 - socket: 12"
+     * "::ffff:99.99.9.999:51165 - socket: 12"
+     * "99.99.9.999:51165 - socket: 12"
+     *
+     * So the IP is the string up the two last colons
+     */
+    int i = g_strlen(rhostname) - 1;
+    while ((i>0) && (rhostname[i] != ':'))
+    {
+        i--;
+    }
+    i--;
+    while ((i>0) && (rhostname[i] != ':'))
+    {
+        i--;
+    }
+
+    hostname = g_strndup(rhostname, i);
+
+    g_memset(&ut, 0, sizeof(ut));
+
+    ut.ut_type=state;
+    ut.ut_pid = pid;
+    gettimeofday(&tv, NULL);
+    ut.ut_tv.tv_sec = tv.tv_sec;
+    ut.ut_tv.tv_usec = tv.tv_usec;
+    g_strncpy(ut.ut_line, line , sizeof(ut.ut_line));
+    g_strncpy(ut.ut_user, user , sizeof(ut.ut_user));
+    g_strncpy(ut.ut_host, hostname, sizeof(ut.ut_host));
+
+    /* utmp */
+    setutxent();
+    pututxline(&ut);
+    endutxent();
+
+    /* wtmp : update on linux, FreeBSD uses utx */
+#ifdef HAVE_UTMPX_H
```

Comment:
i prefer to avoid `ifdef`s inside the function body.
lets add some wrapper to this call outside the function and call it here.
for example:
```c
#if defined(__FreeBSD__)
static inline void
updwtmp(...)
{
...
}
#elif defined(HAVE_UTMP_H)
static inline void
updwtmp(...)
{
...
}
#endif
```